### PR TITLE
[incubator-kie-drools#5705] [new-parser] no viable alternative after …

### PR DESCRIPTION
**Pull request by tkobayas**
…contains operator

**Issue**: 

* https://github.com/apache/incubator-kie-drools/issues/5705

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request and downstream checks</b>  
  - Push a new commit to the PR. An empty commit would be enough.

- for a <b>full downstream build</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- for <b>Jenkins PR check only</b>
  - If you are an ASF committer for KIE podling, login to Jenkins (https://ci-builds.apache.org/job/KIE/job/drools/), go to the specific PR job, and click on `Build Now` button.
</details>

### CODE REVIEW

**Comment by tkobayas** [comment]
@yurloc @gitgabrio @mariofusco Please review, thanks!

**Comment by tkobayas** [comment]
`org.drools.model.codegen.execmodel.CompilerTest#testHashSet` and `org.drools.model.codegen.execmodel.MvelOperatorsTest` passed except another issue (https://github.com/apache/incubator-kie-drools/issues/5708)

Before PR:
`drools-model-codegen`
[ERROR] Tests run: 2433, Failures: 397, Errors: 7, Skipped: 9

After PR:
`drools-model-codegen`
[ERROR] Tests run: 2433, Failures: 370, Errors: 7, Skipped: 9

File: drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRL6Expressions.g4 [hunk 1]
```diff
@@ -838,10 +838,10 @@ in_key
     ;
 
 operator_key
-  // TODO get rid of the DRL_MATCHES token or introduce DRL_CONTAINS etc. for consistency.
-  :      {(helper.isPluggableEvaluator(false))}? id=(IDENTIFIER|DRL_MATCHES) { helper.emit($id, DroolsEditorType.KEYWORD); }
+  // At the moment, we list possible DRLLexer tokens here, but we may be able to improve this by isolating lexers. IDENTIFIER is required to accept custom operators
+  :      {(helper.isPluggableEvaluator(false))}? id=(IDENTIFIER|DRL_MATCHES|DRL_MEMBEROF|DRL_CONTAINS|DRL_EXCLUDES|DRL_SOUNDSLIKE|DRL_STR) { helper.emit($id, DroolsEditorType.KEYWORD); }
```
Comment (gitgabrio) [inline drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRL6Expressions.g4:843]: Thanks @tkobayas !
Could you please explain me what 
` {(helper.isPluggableEvaluator(false))}?` means ?
It seems a condition based on an already `false` condition. Beside, something similar has been removed in a previous [PR](https://github.com/apache/incubator-kie-drools/pull/5759#discussion_r1511345612)

Comment (tkobayas) [reply drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRL6Expressions.g4:843]: Hi @gitgabrio 

`ParserHelper.isPluggableEvaluator` is to check if the next token is a registered operator. The argument is a `boolean` flag to check if the operator is required to support negation.

https://github.com/apache/incubator-kie-drools/blob/dev-new-parser/drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/antlr4/ParserHelper.java#L234

The registry is dynamic so we can add custom operators dynamically.

https://github.com/apache/incubator-kie-drools/blob/dev-new-parser/drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/impl/Operator.java#L102-L106

So we cannot remove the sematic predicate in this case. I updated the PR adding a comment. Thanks.

**Review by mariofusco — APPROVED** [review]

File: drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRL6Expressions.g4 (hunk 1 above)
Comment (gitgabrio) [reply drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRL6Expressions.g4:843]: Thanks @tkobayas ! Maybe this explanation could be written inside ParserHelper itself, sooner or later: wdyt ?

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java [hunk 2]
```diff
@@ -3538,4 +3539,74 @@ void ruleDescrProperties() {
         assertThat(ruleDescr.getName()).isEqualTo("R1");
         assertThat(ruleDescr.getNamespace()).isEqualTo("org.drools");
     }
+
+    /**
+     * Each test input is a constraint expression covering one of the existing keywords. The test is successful if the parser has
+     * no errors and the descriptor's expression string is equal to the input.
+     *
+     * @param constraint expression using a keyword
```
Comment (yurloc) [inline drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java]:
```suggestion
     * @param constraint expression using an operator
```

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java [hunk 3]
```diff
@@ -3538,4 +3539,74 @@ void ruleDescrProperties() {
         assertThat(ruleDescr.getName()).isEqualTo("R1");
         assertThat(ruleDescr.getNamespace()).isEqualTo("org.drools");
     }
+
+    /**
+     * Each test input is a constraint expression covering one of the existing keywords. The test is successful if the parser has
```
Comment (yurloc) [inline drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java]:
```suggestion
     * Each test input is a constraint expression covering one of the existing DRL operators. The test is successful if the parser has
```

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java [hunk 4]
```diff
@@ -3538,4 +3539,74 @@ void ruleDescrProperties() {
...
+            "countries contains \"UK\"",
+            "countries not contains \"UK\"",
+            "countries excludes \"UK\"",
+            "countries not excludes \"UK\"",
+            "firstName soundslike \"John\"",
+            "firstName not soundslike \"John\"",
+            "routingValue str[startsWith] \"R1\"",
+            "routingValue not str[startsWith] \"R1\""
+    })
+    void constraintKeywords(String constraint) {
```
Comment (yurloc) [inline drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java]:
```suggestion
    void constraintOperators(String constraint) {
```

File: drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java [hunk 5]
```diff
@@ -3538,4 +3539,74 @@ void ruleDescrProperties() {
...
+            "countries contains \"UK\" || contains \"US\"",
+            "countries not contains \"UK\" || not contains \"US\"",
+            "countries excludes \"UK\" || excludes \"US\"",
+            "countries not excludes \"UK\" || not excludes \"US\"",
+            "firstName soundslike \"John\" || soundslike \"Paul\"",
+            "firstName not soundslike \"John\" && not soundslike \"Paul\"",
+            "routingValue str[startsWith] \"R1\" || str[startsWith] \"R2\"",
+            "routingValue not str[startsWith] \"R1\" && not str[startsWith] \"R2\""
+    })
+    void halfConstraintKeywords(String constraint) {
```
Comment (yurloc) [inline drools-drl/drools-drl-parser-tests/src/test/java/org/drools/drl/parser/antlr4/MiscDRLParserTest.java]:
```suggestion
    void halfConstraintOperators(String constraint) {
```

**Review by yurloc — APPROVED** [review]
`s/keyword/operator`. Otherwise LGTM

**Comment by tkobayas** [comment]
Thank you, @yurloc !

File: drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRL6Expressions.g4 (hunk 1 above)
Comment (tkobayas) [reply drools-drl/drools-drl-parser/src/main/antlr4/org/drools/drl/parser/antlr4/DRL6Expressions.g4:843]: Sure, added.

**Comment by tkobayas** [comment]
Confirmed that build-chain is fixed so GHA reflects the PR branch.
```
2024-03-07T02:31:54.3357494Z ##[group]Checkout summary
2024-03-07T02:31:54.3358559Z [INFO] [apache/incubator-kie-drools]
2024-03-07T02:31:54.3360248Z [INFO]      Project taken from apache/incubator-kie-drools:dev-new-parser (Branch HEAD 7afbf251a8796cea644b12599b7b236aa2c5fc76)
2024-03-07T02:31:54.3362165Z [INFO]      Merged tkobayas/drools:incubator-kie-drools-5705-contains into branch dev-new-parser
2024-03-07T02:31:54.3363373Z ##[endgroup]
...
2024-03-07T02:36:23.2065568Z [ERROR] Tests run: 52, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 0.969 s <<< FAILURE! -- in org.drools.model.codegen.execmodel.MvelOperatorsTest
2024-03-07T02:36:23.2073196Z [ERROR] org.drools.model.codegen.execmodel.MvelOperatorsTest.testMatchesWithFunction[STANDARD_FROM_DRL] -- Time elapsed: 0.003 s <<< FAILURE!
```

**Review by gitgabrio — APPROVED** [review]
Thanks @tkobayas LGTM!

**Comment by yurloc** [comment]
Fixing progress: There were 457 failed tests (-63).

**Pull request merged**